src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -0,0 +1,421 @@
+  }
+}
+
+int LFUDAPolicy::set_age(int age, optional_yield y) {
+  try {
+    boost::system::error_code ec;
+    response<int> resp;
+    request req;
+    req.push("HSET", "lfuda", "age", std::to_string(age));
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return {};
+
+    return std::get<0>(resp).value(); /* Returns number of fields set */
+  } catch(std::exception &e) {
+    return -1;
+  }
+}
+
+int LFUDAPolicy::get_age(optional_yield y) {
+  response<int> resp;
+
+  try {
+    boost::system::error_code ec;
+    request req;
+    req.push("HEXISTS", "lfuda", "age");
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return -1;
+  } catch(std::exception &e) {
+    return -1;
+  }
+
+  if (!std::get<0>(resp).value()) {
+    if (set_age(0, y)) /* Initialize age */
+      return 0;
+    else
+      return -1;
+  }
+
+  try { 
+    boost::system::error_code ec;
+    response<std::string> value;
+    request req;
+    req.push("HGET", "lfuda", "age");
+      
+    redis_exec(conn, ec, req, value, y);
+
+    if (ec)
+      return -1;
+
+    return std::stoi(std::get<0>(value).value());
+  } catch(std::exception &e) {
+    return -1;
+  }
+}
+
+int LFUDAPolicy::set_local_weight_sum(size_t weight, optional_yield y) {
+  try {
+    boost::system::error_code ec;
+    response<int> resp;
+    request req;
+    req.push("HSET", dir->cct->_conf->rgw_local_cache_address, "localWeights", boost::lexical_cast<int>(weight));
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return -1;
+
+    return std::get<0>(resp).value(); /* Returns number of fields set */
+  } catch(std::exception &e) {
+    return -1;
+  }
+}
+
+int LFUDAPolicy::get_local_weight_sum(optional_yield y) {
+  response<int> resp;
+
+  try {
+    boost::system::error_code ec;
+    request req;
+    req.push("HEXISTS", dir->cct->_conf->rgw_local_cache_address, "localWeights");
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return -1;
+  } catch(std::exception &e) {
+    return -1;
+  }
+
+  if (!std::get<0>(resp).value()) {
+    int sum = 0;
+    for (auto& entry : entries_map)
+      sum += entry.second->localWeight; 
+ 
+    if (set_local_weight_sum(sum, y) < 0) { /* Initialize */ 
+      return -1;
+    } else {
+      return sum;
+    }
+  }
+
+  try { 
+    boost::system::error_code ec;
+    response<std::string> value;
+    request req;
+    req.push("HGET", dir->cct->_conf->rgw_local_cache_address, "localWeights");
+      
+    redis_exec(conn, ec, req, value, y);
+
+    if (ec)
+      return -1;
+
+    return std::stoi(std::get<0>(value).value());
+  } catch(std::exception &e) {
+    return -1;
+  }
+}
+
+CacheBlock* LFUDAPolicy::find_victim(const DoutPrefixProvider* dpp, optional_yield y) {
+  const std::lock_guard l(lfuda_lock);
+  if (entries_heap.empty())
+    return nullptr;
+
+  /* Get victim cache block */
+  std::string key = entries_heap.top()->key;
```

Comment:
@samarahu : I am not sure whether we resolved my comment about two entries/ blocks having the same `weight`?

Reply:
Yes; Orran said that, if this happens, the block is just randomly picked. There is no need for LRU here